pkg/cmd/experimental/buildchain/buildchain.go

```diff
@@ -0,0 +1,333 @@
+				namespaces = append(namespaces, defaultNamespace)
+			}
+
+			// Get all build configurations
+			buildConfigList := make([]buildapi.BuildConfig, 0)
+			for _, namespace := range namespaces {
+				cfgList, err := osc.BuildConfigs(namespace).List(labels.Everything(), labels.Everything())
+				checkErr(err)
+				buildConfigList = append(buildConfigList, cfgList.Items...)
+			}
+
+			// Get all image repositories from build configurations
+			repos := getRepos(buildConfigList, defaultNamespace)
+
+			var tag string
+			if !all && len(args) != 0 {
+				var repo string
+				repo, tag, err = parse(args[0])
+				checkErr(err)
+
+				// Validate the specified image repository
+				repo = join(defaultNamespace, repo)
+				if _, ok := repos[repo]; !ok {
+					checkErr(fmt.Errorf("invalid image repository: %s", repo))
+				}
+
+				// Flush the map
+				repos = map[string]struct{}{}
+				// Set the specified repo as the only one to output dependencies for
+				repos[repo] = struct{}{}
+			} else {
+				// Make sure from now on that the --all flag is true
+				// since we are building dependency trees for every
+				// image repository available either in the current
+				// namespace or in every namespace
+				all = true
+			}
+
+			if len(repos) == 0 {
+				checkErr(fmt.Errorf("no image repository available for building its dependency tree"))
+			}
+
+			output := cmdutil.GetFlagString(cmd, "output")
+			for repo := range repos {
+				root := findRepoDeps(repo, tag, all, buildConfigList)
+
+				switch output {
+				case "json":
+					jsonDump, err := json.MarshalIndent(root, "", "\t")
+					checkErr(err)
+					fmt.Println(string(jsonDump))
+				case "dot":
+					g := dot.NewGraph()
+					// The DOT parser yells on names with dashes
+					_, name := split(repo)
+					graphName := validDOT(name)
+					g.SetName(graphName)
+					// Directed graph since we illustrate dependencies
+					g.SetDir(true)
+					// Explicitly allow multiple pairs of edges between
+					// the same pair of nodes
+					g.SetStrict(false)
+					fmt.Println(dotDump(root, g, graphName))
+				case "ast":
+					fmt.Println(root)
+				default:
+					usageError(cmd, "Wrong output format specified: %s", output)
+				}
+			}
+		},
+	}
+
+	cmd.Flags().Bool("all", false, "Build dependency trees for all image repositories")
+	cmd.Flags().StringP("output", "o", "json", "Output format of dependency tree(s)")
+	return cmd
+}
+
+// getRepos iterates over a given set of build configurations
+// and extracts all the image repositories on which other
+// image repositories are based
+func getRepos(configs []buildapi.BuildConfig, defaultNamespace string) map[string]struct{} {
+	avoidDuplicates := make(map[string]struct{})
+	var repo string
+	for _, cfg := range configs {
+		for _, tr := range cfg.Triggers {
+			if tr.ImageChange != nil && tr.ImageChange.From.Name != "" {
+				switch tr.ImageChange.From.Namespace {
+				case "":
+					repo = join(defaultNamespace, tr.ImageChange.From.Name)
```

Comment:
Is defaultNamespace the right namespace to assign here or should I check first if cfg.Namespace is defined, if so, assign that, if not, assign defaultNamespace ? @bparees